Source/missiles.cpp

```diff
@@ -410,19 +416,24 @@ void CheckMissileCol(Missile &missile, int minDamage, int maxDamage, bool isDama
 
 	bool isPlayerHit = false;
 	bool blocked = false;
-	const int8_t pid = dPlayer[mx][my];
+	int8_t pid = dPlayer[mx][my];
 	if (pid > 0) {
-		if (missile._micaster != TARGET_BOTH && !missile.IsTrap()) {
-			if (missile._micaster == TARGET_MONSTERS) {
-				if ((pid - 1) != missile._misource)
-					isPlayerHit = Plr2PlrMHit(Players[missile._misource], pid - 1, minDamage, maxDamage, missile._midist, missile._mitype, isDamageShifted, &blocked);
-			} else {
-				Monster &monster = Monsters[missile._misource];
-				isPlayerHit = PlayerMHit(pid - 1, &monster, missile._midist, minDamage, maxDamage, missile._mitype, isDamageShifted, 0, &blocked);
+		pid -= 1;
+		int earflag = 0;
+		switch (missileSource) {
+		case MissileSource::Player: {
+			Player *player = missile.sourcePlayer();
+			if (&Players[pid] != player) {
+				isPlayerHit = Plr2PlrMHit(*player, pid, minDamage, maxDamage, missile._midist, missile._mitype, isDamageShifted, &blocked);
+				break;
 			}
-		} else {
-			int earflag = (!missile.IsTrap() && (missile._miAnimType == MFILE_FIREWAL || missile._miAnimType == MFILE_LGHNING)) ? 1 : 0;
-			isPlayerHit = PlayerMHit(pid - 1, nullptr, missile._midist, minDamage, maxDamage, missile._mitype, isDamageShifted, earflag, &blocked);
+			earflag = IsAnyOf(missile._miAnimType, MFILE_FIREWAL, MFILE_LGHNING) ? 1 : 0;

```

Comment:
There is a bug in the original Hellfire here. The play will not drop an ear if killed by a fire rune, but will for lightning runes.

Reply:
I don't know if you have seen the original 200 line monstrosity: https://github.com/diasurgical/devilution/blob/b86b8fd7c4966385e3a7dbcb29e6d7e05bdab95e/Source/missiles.cpp#L1131